File: backend/taskwarrior/sortspec.go
package taskwarrior

import (
	"sort"
	"strings"
)

// defaultSortSpec is used when the configuration carries no sort order for
// the report, matching Taskwarrior's `next` report.
const defaultSortSpec = "urgency-"

// sortReport is the report whose sort order the task list follows. `list` is
// the report `task list` uses, so the UI and the command line agree.
const sortReport = "list"

// SortSpec describes the ordering applied to a task list.
type SortSpec struct {
	// Report names the Taskwarrior report the order was taken from.
	Report string `json:"report"`
	// Spec is the raw configuration value, e.g. "start-,due+,project+,urgency-".
	Spec string `json:"spec"`
	// Unsupported lists attributes teutates cannot sort by; they are skipped
	// rather than silently changing the order.
	Unsupported []string `json:"unsupported,omitempty"`
}

// sortKey is one clause of a sort specification.
type sortKey struct {
	attribute  string
	descending bool
}

// attributeKind determines how missing values are treated.
type attributeKind int

const (
	// kindDate covers timestamps. Taskwarrior sorts tasks without the date
	// last in both directions, so an undated task never leads the list.
	kindDate attributeKind = iota
	// kindText compares lexicographically, which puts an empty value first
	// ascending and last descending — how `project+` behaves.
	kindText
	// kindNumber treats a missing value as zero.
	kindNumber
	// kindPriority is an ordered enum where "no priority" ranks below L.
	kindPriority
)

// sortableAttributes maps the attribute names a sort spec may use onto the
// task field they read. Anything absent here is reported as unsupported.
var sortableAttributes = map[string]struct {
	kind attributeKind
	text func(Task) string
	num  func(Task) float64
}{
	"due":         {kind: kindDate, text: func(t Task) string { return t.Due }},
	"start":       {kind: kindDate, text: func(t Task) string { return t.Start }},
	"end":         {kind: kindDate, text: func(t Task) string { return t.End }},
	"entry":       {kind: kindDate, text: func(t Task) string { return t.Entry }},
	"modified":    {kind: kindDate, text: func(t Task) string { return t.Modified }},
	"scheduled":   {kind: kindDate, text: func(t Task) string { return t.Scheduled }},
	"wait":        {kind: kindDate, text: func(t Task) string { return t.Wait }},
	"project":     {kind: kindText, text: func(t Task) string { return t.Project }},
	"description": {kind: kindText, text: func(t Task) string { return t.Description }},
	"status":      {kind: kindText, text: func(t Task) string { return t.Status }},
	"uuid":        {kind: kindText, text: func(t Task) string { return t.UUID }},
	"urgency":     {kind: kindNumber, num: func(t Task) float64 { return t.Urgency }},
	"id":          {kind: kindNumber, num: func(t Task) float64 { return float64(t.ID) }},
	"priority":    {kind: kindPriority, text: func(t Task) string { return t.Priority }},
}

// priorityRank orders priorities. A task without a priority ranks lowest, so
// `priority-` lists H, M, L and then the unprioritised ones.
var priorityRank = map[string]int{"L": 1, "M": 2, "H": 3}

// parseSortSpec reads a Taskwarrior sort specification such as
// "start-,due+,project+,urgency-". A trailing "/" marks a visual break in
// Taskwarrior's reports and has no effect on the order, so it is stripped.
func parseSortSpec(spec string) ([]sortKey, []string) {
	var keys []sortKey
	var unsupported []string

	for _, clause := range strings.Split(spec, ",") {
		clause = strings.TrimSpace(clause)
		clause = strings.TrimSuffix(clause, "/")
		if clause == "" {
			continue
		}

		descending := false
		switch clause[len(clause)-1] {
		case '-':
			descending = true
			clause = clause[:len(clause)-1]
		case '+':
			clause = clause[:len(clause)-1]
		}
		if clause == "" {
			continue
		}

		if _, ok := sortableAttributes[clause]; !ok {
			unsupported = append(unsupported, clause)
			continue
		}
		keys = append(keys, sortKey{attribute: clause, descending: descending})
	}

	return keys, unsupported
}

// compareBy returns -1, 0 or 1 for a single sort clause.
func compareBy(a, b Task, key sortKey) int {
	attribute := sortableAttributes[key.attribute]

	var result int
	switch attribute.kind {
	case kindDate:
		left, right := attribute.text(a), attribute.text(b)
		// A missing date sorts last whichever way the clause points, so the
		// comparison is returned before the direction is applied.
		switch {
		case left == "" && right == "":
			return 0
		case left == "":
			return 1
		case right == "":
			return -1
		}
		// Dates are RFC 3339 in UTC and therefore fixed width, which makes
		// lexicographic order the same as chronological order.
		result = strings.Compare(left, right)

	case kindPriority:
		left, right := priorityRank[attribute.text(a)], priorityRank[attribute.text(b)]
		switch {
		case left < right:
			result = -1
		case left > right:
			result = 1
		}

	case kindNumber:
		left, right := attribute.num(a), attribute.num(b)
		switch {
		case left < right:
			result = -1
		case left > right:
			result = 1
		}

	default:
		result = strings.Compare(attribute.text(a), attribute.text(b))
	}

	if key.descending {
		return -result
	}
	return result
}

// sortTasksBy orders tasks by the given clauses. Ties fall back to the
// working id and then the uuid, so the order is stable across requests even
// for completed tasks, which all export an id of 0.
func sortTasksBy(tasks []Task, keys []sortKey) {
	sort.SliceStable(tasks, func(i, j int) bool {
		for _, key := range keys {
			if result := compareBy(tasks[i], tasks[j], key); result != 0 {
				return result < 0
			}
		}
		if tasks[i].ID != tasks[j].ID {
			return tasks[i].ID < tasks[j].ID
		}
		return tasks[i].UUID < tasks[j].UUID
	})
}

// sortSpecFor reads the sort order configured for the report teutates
// follows, falling back to the default when it is not configured.
func sortSpecFor(config *Config) string {
	if config == nil {
		return defaultSortSpec
	}
	key := "report." + sortReport + ".sort"
	for _, setting := range config.Settings {
		if setting.Key == key && setting.Value != "" {
			return setting.Value
		}
	}
	return defaultSortSpec
}

File: ui/src/taskwarrior/sortspec.ts
import type { Config } from "./config";
import type { Task } from "./tasks";

/** Used when the configuration carries no sort order for the report,
 *  matching Taskwarrior's `next` report. */
const defaultSortSpec = "urgency-";

/** The report whose sort order the task list follows. `list` is the report
 *  `task list` uses, so the interface and the command line agree. */
export const sortReport = "list";

/** Describes the ordering applied to a task list. */
export interface SortSpec {
  /** The Taskwarrior report the order was taken from, e.g. "list". */
  report: string;
  /** The raw configuration value, e.g. "start-,due+,project+,urgency-". */
  spec: string;
  /** Attributes teutates cannot sort by; they are skipped rather than
   *  silently changing the order. */
  unsupported?: string[];
}

/** One clause of a sort specification. */
export interface SortKey {
  attribute: string;
  descending: boolean;
}

/** How missing values are treated.
 *
 *  - `date`: Taskwarrior sorts tasks without the date last in both
 *    directions, so an undated task never leads the list.
 *  - `text`: compares lexicographically, which puts an empty value first
 *    ascending and last descending — how `project+` behaves.
 *  - `number`: treats a missing value as zero.
 *  - `priority`: an ordered enum where "no priority" ranks below L. */
type AttributeKind = "date" | "text" | "number" | "priority";

/** An attribute and how to read the field it compares. Numeric attributes
 *  read a number; every other kind reads text. */
type Attribute =
  | { kind: Exclude<AttributeKind, "number">; read: (task: Task) => string }
  | { kind: "number"; read: (task: Task) => number };

/** Maps the attribute names a sort spec may use onto the task field they
 *  read. Anything absent here is reported as unsupported. */
const sortableAttributes: Record<string, Attribute> = {
  due: { kind: "date", read: (t) => t.due ?? "" },
  start: { kind: "date", read: (t) => t.start ?? "" },
  end: { kind: "date", read: (t) => t.end ?? "" },
  entry: { kind: "date", read: (t) => t.entry ?? "" },
  modified: { kind: "date", read: (t) => t.modified ?? "" },
  scheduled: { kind: "date", read: (t) => t.scheduled ?? "" },
  wait: { kind: "date", read: (t) => t.wait ?? "" },
  project: { kind: "text", read: (t) => t.project ?? "" },
  description: { kind: "text", read: (t) => t.description },
  status: { kind: "text", read: (t) => t.status },
  uuid: { kind: "text", read: (t) => t.uuid },
  urgency: { kind: "number", read: (t) => t.urgency },
  id: { kind: "number", read: (t) => t.id },
  priority: { kind: "priority", read: (t) => t.priority ?? "" },
};

/** Orders priorities. A task without a priority ranks lowest, so `priority-`
 *  lists H, M, L and then the unprioritised ones. */
const priorityRank: Record<string, number> = { L: 1, M: 2, H: 3 };

function compareText(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

/** Reads a Taskwarrior sort specification such as
 *  "start-,due+,project+,urgency-". A trailing "/" marks a visual break in
 *  Taskwarrior's reports and has no effect on the order, so it is stripped. */
export function parseSortSpec(spec: string): {
  keys: SortKey[];
  unsupported: string[];
} {
  const keys: SortKey[] = [];
  const unsupported: string[] = [];

  for (const raw of spec.split(",")) {
    let clause = raw.trim().replace(/\/$/, "");
    if (clause === "") continue;

    let descending = false;
    const last = clause[clause.length - 1];
    if (last === "-") {
      descending = true;
      clause = clause.slice(0, -1);
    } else if (last === "+") {
      clause = clause.slice(0, -1);
    }
    if (clause === "") continue;

    if (!Object.hasOwn(sortableAttributes, clause)) {
      unsupported.push(clause);
      continue;
    }
    keys.push({ attribute: clause, descending });
  }

  return { keys, unsupported };
}

/** Returns -1, 0 or 1 for a single sort clause. */
export function compareBy(a: Task, b: Task, key: SortKey): number {
  const attribute = sortableAttributes[key.attribute];
  // parseSortSpec drops clauses naming an attribute that is not here, so
  // this only guards a key assembled some other way.
  if (attribute === undefined) return 0;

  let result = 0;
  switch (attribute.kind) {
    case "date": {
      const left = attribute.read(a);
      const right = attribute.read(b);
      // A missing date sorts last whichever way the clause points, so the
      // comparison is returned before the direction is applied.
      if (left === "" && right === "") return 0;
      if (left === "") return 1;
      if (right === "") return -1;
      // Dates are RFC 3339 in UTC and therefore fixed width, which makes
      // lexicographic order the same as chronological order.
      result = compareText(left, right);
      break;
    }

    case "priority": {
      const left = priorityRank[attribute.read(a)] ?? 0;
      const right = priorityRank[attribute.read(b)] ?? 0;
      result = left < right ? -1 : left > right ? 1 : 0;
      break;
    }

    case "number": {
      const left = attribute.read(a);
      const right = attribute.read(b);
      result = left < right ? -1 : left > right ? 1 : 0;
      break;
    }

    default:
      result = compareText(attribute.read(a), attribute.read(b));
  }

  return key.descending ? -result : result;
}

/** Orders tasks by the given clauses. Ties fall back to the working id and
 *  then the uuid, so the order is stable across requests even for completed
 *  tasks, which all export an id of 0. */
export function sortTasksBy(tasks: Task[], keys: SortKey[]): void {
  tasks.sort((a, b) => {
    for (const key of keys) {
      const result = compareBy(a, b, key);
      if (result !== 0) return result;
    }
    if (a.id !== b.id) return a.id - b.id;
    return compareText(a.uuid, b.uuid);
  });
}

/** Reads the sort order configured for the report teutates follows, falling
 *  back to the default when it is not configured. */
export function sortSpecFor(config: Config | undefined): string {
  if (config === undefined) return defaultSortSpec;

  const key = `report.${sortReport}.sort`;
  for (const setting of config.settings) {
    if (setting.key === key && setting.value !== "") return setting.value;
  }
  return defaultSortSpec;
}
